### ISSUE TITLE
[Issue]: 7900 xtx producing NaN values when training and/or provides bad results.

### ISSUE THREAD

**Issue by CrazyCows** [ROCm/ROCm#4391]
### Problem Description

I had two issues when training. Both models is pulled from HF. Full training passes has been succesful using nvidia GPU's, and partial training (reduced dataset) has been succesful using apple ARM chips - neither of which suceeded on 7900 XTX.

Issues:
1. NaN values start appearing after a few thousand steps when fine tuning bert from google, specifically "google/bert-base-multilingual-uncased".
2. Accuracy is very, very poor when training succeeds on fine tuning facebooks model roberta-base, specifically "roberta-base".  When fine-tuning on apple/nvidia, we reach ~98% accuracy. on amd we reach ~35% accuracy, clearly denoting an issue.

Furthermore, both bert and roberta produces NaN values when enabling mixed precision. Bert produces NaN values regardless of precision.

The original script which is provided below uses HF wrappers. Script has been rewritten using only torch and transformers in an attempt to improve results, but to no avail.

### Operating System

24.04.01 LTS

### CPU

ryzen 9700x

### GPU

7900 xtx

### ROCm Version

ROCm 6.3.2

### Steps to Reproduce

1. Download radeon driver and ROCm from [amds official site](https://www.amd.com/en/support/download/linux-drivers.html)
2. Instantiate venv using python 3.11.9
3. Install dependencies: 
`sentence-transformers = "^3.3.1"
sentencepiece = "^0.2.0"`
4. pip install torch. Will uninstall torch from sentence-transformers and install correct version - "pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/rocm6.2.4"
5. Add some data. My dataset is propeitary and written in danish. Thus special chars "åæø" is prevelant. Token size for all entries is <100 with ~2 million entries.  

`from sentence_transformers import CrossEncoder, InputExample
from torch.utils.data import DataLoader
from sentence_transformers.cross_encoder.evaluation import CEBinaryClassificationEvaluator
import math
import os
from sklearn.model_selection import train_test_split

os.environ["CUDA_VISIBLE_DEVICES"] = "0"  # To ensure onboard graphics is not chosen.

def some_cross_encoder(X_train, y_train, X_val, y_val, epochs=5, batch_size=32, learning_rate=2e-5, model_folder_name="version"):
    model = CrossEncoder('google/bert-base-multilingual-uncased', num_labels=1)

    # model.model.to(torch.float32)
    device = "cuda" if torch.cuda.is_available() else "cpu" 
    model.model.to(device)

    train_examples = [InputExample(texts=[sent1, sent2], label=label) for (sent1, sent2), label in zip(X_train, y_train)]
    val_examples = [InputExample(texts=[sent1, sent2], label=label) for (sent1, sent2), label in zip(X_val, y_val)]
    
    train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=batch_size)
    evaluator = CEBinaryClassificationEvaluator.from_input_examples(val_examples, name="Evaluate dataset")

    warmup_steps = math.ceil(len(train_dataloader) * epochs * 0.1) 
    model.fit(
        train_dataloader=train_dataloader,
        evaluator=evaluator,
        epochs=epochs,
        warmup_steps=warmup_steps,
        optimizer_params={'lr': learning_rate},
        save_best_model=True,
        output_path=(model_folder_name + '_best'),
    )

    model.save(model_folder_name)

    return model

# X: [(str, str)], y: [float]
X_t, X_v, y_t, y_v = train_test_split(X, y, test_size=0.1, random_state=42)
some_cross_encoder(X_t, X_v, y_t, y_v)
`

**Comment by ruizhe-ops** [comment]
In your current code,you are using 32-bit floating-point precision(torch.float32).You can switch to Mixed Precision Training(MPT),which combines FP32 and FP16 precision to improve training speed and reduce memory usage while avoiding numerical instability issues.PyTorch provides the torch.cuda.amp module to support mixed precision training.

Below is the modified code to enable mixed precision training:

from sentence_transformers import CrossEncoder, InputExample
from torch.utils.data import DataLoader
from sentence_transformers.cross_encoder.evaluation import CEBinaryClassificationEvaluator
import math
import os
import torch
from sklearn.model_selection import train_test_split
from torch.cuda.amp import GradScaler, autocast  # Import mixed precision tools

os.environ["CUDA_VISIBLE_DEVICES"] = "0"  # Ensure the integrated graphics card is not used

def some_cross_encoder(X_train, y_train, X_val, y_val, epochs=5, batch_size=32, learning_rate=2e-5, model_folder_name="version"):
    model = CrossEncoder('google/bert-base-multilingual-uncased', num_labels=1)

    device = "cuda" if torch.cuda.is_available() else "cpu"
    model.model.to(device)

    train_examples = [InputExample(texts=[sent1, sent2], label=label) for (sent1, sent2), label in zip(X_train, y_train)]
    val_examples = [InputExample(texts=[sent1, sent2], label=label) for (sent1, sent2), label in zip(X_val, y_val)]
    
    train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=batch_size)
    evaluator = CEBinaryClassificationEvaluator.from_input_examples(val_examples, name="Evaluate dataset")

    warmup_steps = math.ceil(len(train_dataloader) * epochs * 0.1)

    # Initialize GradScaler
    scaler = GradScaler()

    # Modify the fit method to support mixed precision training
    def train_step(model, inputs, labels, optimizer, scaler):
        model.train()
        with autocast():  # Enable mixed precision
            outputs = model(inputs)
            loss = outputs.loss
        scaler.scale(loss).backward()  # Scale the loss and perform backpropagation
        scaler.step(optimizer)  # Update the weights
        scaler.update()  # Update the scaling factor
        model.zero_grad()

    # Initialize the optimizer
    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)

    for epoch in range(epochs):
        for batch in train_dataloader:
            inputs = batch.texts
            labels = batch.label
            train_step(model, inputs, labels, optimizer, scaler)

        # Evaluate the model
        evaluator(model)

    model.save(model_folder_name + "_best")
    model.save(model_folder_name)

    return model

# Example data
X = [("This is a test sentence.", "This is another test sentence."), ("Another pair.", "Of sentences.")]
y = [1.0, 0.0]

# Split the data
X_t, X_v, y_t, y_v = train_test_split(X, y, test_size=0.1, random_state=42)

# Train the model
some_cross_encoder(X_t, X_v, y_t, y_v)

Additionally,the link you provided points to ROCm 6.2.4,but your system is using ROCm 6.3.2.This may cause compatibility issues.Since PyTorch may already include ROCm,you can try installing PyTorch directly without installing the official ROCm.If the problem persists,it is recommended to submit an Issue on the PyTorch community forum with detailed environment information.

**Comment by ruizhe-ops** [comment]
In your current code,you are using 32-bit floating-point precision(torch.float32).You can switch to Mixed Precision Training(MPT),which combines FP32 and FP16 precision to improve training speed and reduce memory usage while avoiding numerical instability issues.PyTorch provides the torch.cuda.amp module to support mixed precision training.

Below is the modified code to enable mixed precision training:

from sentence_transformers import CrossEncoder, InputExample
from torch.utils.data import DataLoader
from sentence_transformers.cross_encoder.evaluation import CEBinaryClassificationEvaluator
import math
import os
import torch
from sklearn.model_selection import train_test_split
from torch.cuda.amp import GradScaler, autocast  # Import mixed precision tools

os.environ["CUDA_VISIBLE_DEVICES"] = "0"  # Ensure the integrated graphics card is not used

def some_cross_encoder(X_train, y_train, X_val, y_val, epochs=5, batch_size=32, learning_rate=2e-5, model_folder_name="version"):
    model = CrossEncoder('google/bert-base-multilingual-uncased', num_labels=1)

    device = "cuda" if torch.cuda.is_available() else "cpu"
    model.model.to(device)

    train_examples = [InputExample(texts=[sent1, sent2], label=label) for (sent1, sent2), label in zip(X_train, y_train)]
    val_examples = [InputExample(texts=[sent1, sent2], label=label) for (sent1, sent2), label in zip(X_val, y_val)]
    
    train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=batch_size)
    evaluator = CEBinaryClassificationEvaluator.from_input_examples(val_examples, name="Evaluate dataset")

    warmup_steps = math.ceil(len(train_dataloader) * epochs * 0.1)

    # Initialize GradScaler
    scaler = GradScaler()

    # Modify the fit method to support mixed precision training
    def train_step(model, inputs, labels, optimizer, scaler):
        model.train()
        with autocast():  # Enable mixed precision
            outputs = model(inputs)
            loss = outputs.loss
        scaler.scale(loss).backward()  # Scale the loss and perform backpropagation
        scaler.step(optimizer)  # Update the weights
        scaler.update()  # Update the scaling factor
        model.zero_grad()

    # Initialize the optimizer
    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)

    for epoch in range(epochs):
        for batch in train_dataloader:
            inputs = batch.texts
            labels = batch.label
            train_step(model, inputs, labels, optimizer, scaler)

        # Evaluate the model
        evaluator(model)

    model.save(model_folder_name + "_best")
    model.save(model_folder_name)

    return model

# Example data
X = [("This is a test sentence.", "This is another test sentence."), ("Another pair.", "Of sentences.")]
y = [1.0, 0.0]

# Split the data
X_t, X_v, y_t, y_v = train_test_split(X, y, test_size=0.1, random_state=42)

# Train the model
some_cross_encoder(X_t, X_v, y_t, y_v)

Additionally,the link you provided points to ROCm 6.2.4,but your system is using ROCm 6.3.2.This may cause compatibility issues.Since PyTorch may already include ROCm,you can try installing PyTorch directly without installing the official ROCm.If the problem persists,it is recommended to submit an Issue on the PyTorch community forum with detailed environment information.

**Comment by ruizhe-ops** [comment]
In your current code,you are using 32-bit floating-point precision(torch.float32).You can switch to Mixed Precision Training(MPT),which combines FP32 and FP16 precision to improve training speed and reduce memory usage while avoiding numerical instability issues.PyTorch provides the torch.cuda.amp module to support mixed precision training.

Below is the modified code to enable mixed precision training:

from sentence_transformers import CrossEncoder, InputExample
from torch.utils.data import DataLoader
from sentence_transformers.cross_encoder.evaluation import CEBinaryClassificationEvaluator
import math
import os
import torch
from sklearn.model_selection import train_test_split
from torch.cuda.amp import GradScaler, autocast  # Import mixed precision tools

os.environ["CUDA_VISIBLE_DEVICES"] = "0"  # Ensure the integrated graphics card is not used

def some_cross_encoder(X_train, y_train, X_val, y_val, epochs=5, batch_size=32, learning_rate=2e-5, model_folder_name="version"):
    model = CrossEncoder('google/bert-base-multilingual-uncased', num_labels=1)

    device = "cuda" if torch.cuda.is_available() else "cpu"
    model.model.to(device)

    train_examples = [InputExample(texts=[sent1, sent2], label=label) for (sent1, sent2), label in zip(X_train, y_train)]
    val_examples = [InputExample(texts=[sent1, sent2], label=label) for (sent1, sent2), label in zip(X_val, y_val)]
    
    train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=batch_size)
    evaluator = CEBinaryClassificationEvaluator.from_input_examples(val_examples, name="Evaluate dataset")

    warmup_steps = math.ceil(len(train_dataloader) * epochs * 0.1)

    # Initialize GradScaler
    scaler = GradScaler()

    # Modify the fit method to support mixed precision training
    def train_step(model, inputs, labels, optimizer, scaler):
        model.train()
        with autocast():  # Enable mixed precision
            outputs = model(inputs)
            loss = outputs.loss
        scaler.scale(loss).backward()  # Scale the loss and perform backpropagation
        scaler.step(optimizer)  # Update the weights
        scaler.update()  # Update the scaling factor
        model.zero_grad()

    # Initialize the optimizer
    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)

    for epoch in range(epochs):
        for batch in train_dataloader:
            inputs = batch.texts
            labels = batch.label
            train_step(model, inputs, labels, optimizer, scaler)

        # Evaluate the model
        evaluator(model)

    model.save(model_folder_name + "_best")
    model.save(model_folder_name)

    return model

X = [("This is a test sentence.", "This is another test sentence."), ("Another pair.", "Of sentences.")]
y = [1.0, 0.0]

X_t, X_v, y_t, y_v = train_test_split(X, y, test_size=0.1, random_state=42)

some_cross_encoder(X_t, X_v, y_t, y_v)

Additionally,the link you provided points to ROCm 6.2.4,but your system is using ROCm 6.3.2.This may cause compatibility issues.Since PyTorch may already include ROCm,you can try installing PyTorch directly without installing the official ROCm.If the problem persists,it is recommended to submit an Issue on the PyTorch community forum with detailed environment information.

**Comment by ruizhe-ops** [comment]
In your current code,you are using 32-bit floating-point precision(torch.float32).You can switch to Mixed Precision Training(MPT),which combines FP32 and FP16 precision to improve training speed and reduce memory usage while avoiding numerical instability issues.PyTorch provides the torch.cuda.amp module to support mixed precision training.
Below is the modified code to enable mixed precision training:

from sentence_transformers import CrossEncoder, InputExample
from torch.utils.data import DataLoader
from sentence_transformers.cross_encoder.evaluation import CEBinaryClassificationEvaluator
import math
import os
import torch
from sklearn.model_selection import train_test_split
from torch.cuda.amp import GradScaler, autocast  # Import mixed precision tools

os.environ["CUDA_VISIBLE_DEVICES"] = "0"  # Ensure the integrated graphics card is not used

def some_cross_encoder(X_train, y_train, X_val, y_val, epochs=5, batch_size=32, learning_rate=2e-5, 
model_folder_name="version"):

    model = CrossEncoder('google/bert-base-multilingual-uncased', num_labels=1)
    device = "cuda" if torch.cuda.is_available() else "cpu"
    model.model.to(device)
    train_examples = [InputExample(texts=[sent1, sent2], label=label) for (sent1, sent2), label in zip(X_train, y_train)]
    val_examples = [InputExample(texts=[sent1, sent2], label=label) for (sent1, sent2), label in zip(X_val, y_val)]
    train_dataloader = DataLoader(train_examples, shuffle=True, batch_size=batch_size)
    evaluator = CEBinaryClassificationEvaluator.from_input_examples(val_examples, name="Evaluate dataset")
    warmup_steps = math.ceil(len(train_dataloader) * epochs * 0.1)
    # Initialize GradScaler
    scaler = GradScaler()
    # Modify the fit method to support mixed precision training
    def train_step(model, inputs, labels, optimizer, scaler):
        model.train()
        with autocast():  # Enable mixed precision
            outputs = model(inputs)
            loss = outputs.loss
        scaler.scale(loss).backward()  # Scale the loss and perform backpropagation
        scaler.step(optimizer)  # Update the weights
        scaler.update()  # Update the scaling factor
        model.zero_grad()
    optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)
    for epoch in range(epochs):
        for batch in train_dataloader:
            inputs = batch.texts
            labels = batch.label
            train_step(model, inputs, labels, optimizer, scaler)
        evaluator(model)
    model.save(model_folder_name + "_best")
    model.save(model_folder_name)
    return model

X = [("This is a test sentence.", "This is another test sentence."), ("Another pair.", "Of sentences.")]
y = [1.0, 0.0]
X_t, X_v, y_t, y_v = train_test_split(X, y, test_size=0.1, random_state=42)
some_cross_encoder(X_t, X_v, y_t, y_v)

Additionally,the link you provided points to ROCm 6.2.4,but your system is using ROCm 6.3.2.This may cause compatibility issues.Since PyTorch may already include ROCm,you can try installing PyTorch directly without installing the official ROCm.If the problem persists,it is recommended to submit an Issue on the PyTorch community forum with detailed environment information.

**Comment by powderluv** [comment]
Here is a version to try: https://gist.github.com/powderluv/f7e62938209dc3479ca86553df608bc1

**Comment by ruizhe-ops** [comment]
This issue may stem from the ROCm framework's incomplete support for mixed precision (FP16/FP32) on AMD GPUs, leading to numerical instability, or it could be due to uncovered operators in the integration between PyTorch and ROCm. And another possibility is that the AI workload exceeds the performance limits of the Radeon RX 7900 XTX. 
By the way, I’m not sure which NVIDIA GPU model you are using. To further troubleshoot, you might consider checking the GPU's temperature, frequency, and memory utilization to rule out any thermal or power-related issues.

**Comment by pssiva152** [comment]
Hi @powderluv, The reported issue is not reproduced in small input data however when we increase the input data we hit the NaN issue in both BERT and RoBERTa models. I created this repo (https://github.com/pssiva152/debug) to repro the issue and I'm able to repo it with **rocm/pytorch:latest docker** but with **rocm/pytorch-nightly:latest** docker issue is not seen. Do you suggest having an internal tracking ticket to track this issue?

**Comment by CrazyCows** [comment]
I got my hands on another card. As denoted by @pssiva152, the nightly docker container does indeed fix the issue for me as well. Furthermore, using the latest driver and the nightly torch build without a docker container also solves the issue. Nightly works and produces good results with both mixed precision and full fp32. 

However, on a slightly unrelated note the nightly build does cut the training speed in half - archiving speeds slightly faster than a 2070 super, which is unimpressive but ey, it works and wasn't totally unexpected. 

Someone else asked what cards it has otherwise been tested on:

2070 super
4070 ti
L4
A10
H100
(honorable mention, apple m1)

**Comment by ppanchad-amd** [comment]
Hi @CrazyCows. Closing ticket since nightly docker container has the fix. Thanks!
